src/contexts/loginContext.tsx

```diff
@@ -31,18 +35,50 @@ export const LoginProvider = ({ children }: ProviderProps) => {
     }
   }, [user, authLoading])
 
+  const withTokenRefresh = (fn: ApiCall) => {
+    if (!user || !token) return
+
+    // Set currentToken inside the closure because when this
+    // is called inside a callback, the token won't update while
+    // the function runs.
+    let currentToken = token
+    let retryAttempt = 0
+
+    while (retryAttempt < 2) {
+      retryAttempt += 1
+
+      try {
+        fn(currentToken)
+        break
+      } catch (e: any) {
+        if (retryAttempt < 2) {
+          if (import.meta.env.DEV)
+            console.log('Server returned status 401, retrying...')
+
+          user?.getIdToken(true).then((idToken) => {
+            setToken(idToken)
+            currentToken = idToken
+          })
+        } else {
+          signOutWithGoogle()
+        }
+      }
+    }
+  }
+
   const value = {
     user,
     token,
-    requireLogin,
     authLoading,
+    requireLogin,
+    withTokenRefresh,
   }
 
   useEffect(() => {
     if (authError) signOutWithGoogle()
 
     if (user) {
-      user.getIdToken(true).then((token) => setToken(token))
+      user.getIdToken(true).then((idToken) => setToken(idToken))
```

Comment:
Not strictly related to this change, but we want to eliminate "shadowed" variables, if possible.